gnucash/import-export/csv-imp/gnc-csv-price-import-settings.cpp

```diff
@@ -0,0 +1,258 @@
+
+#define CSV_COL_TYPES    "ColumnTypes"
+
+#define CSV_TO_CURR      "PriceToCurrency"
+#define CSV_FROM_COMM    "PriceFromCommodity"
+
+G_GNUC_UNUSED static QofLogModule log_module = GNC_MOD_IMPORT;
+
+preset_vec_price presets_price;
+
+static std::shared_ptr<CsvPriceImpSettings> create_int_no_preset(void)
+{
+    auto preset = std::make_shared<CsvPriceImpSettings>();
+    preset->m_name = get_no_settings();
+    preset->m_settings_type = settings_type;
+
+    return preset;
+}
+
+static std::shared_ptr<CsvPriceImpSettings> create_int_gnc_exp_preset(void)
+{
+    auto preset = std::make_shared<CsvPriceImpSettings>();
+    preset->m_name = get_gnc_exp();
+    preset->m_skip_start_lines = 1;
+
+    /* FIXME date and currency format should still be aligned with export format!
+     * That's currently hard to do, because the export uses whatever the user
+     * had set as global preference.
+    preset->date_active = 0;
+    preset->currency_active = 0;
+    */
+    preset->m_column_types_price = {
+            GncPricePropType::DATE,
+            GncPricePropType::AMOUNT,
+            GncPricePropType::FROM_COMMODITY,
+            GncPricePropType::TO_CURRENCY
+    };
+    return preset;
+}
+
+/**************************************************
+ * find
+ *
+ * find all settings entries in the state key file
+ * based on settings type.
+ **************************************************/
+const preset_vec_price& get_import_presets_price (void)
+{
+    // Search all Groups in the state key file for ones starting with prefix
+    auto preset_names = std::vector<std::string>();
+    auto keyfile = gnc_state_get_current ();
+    gsize grouplength;
+    gchar **groups = g_key_file_get_groups (keyfile, &grouplength);
+
+    /* Start by building a sorted list of candidate presets as found in the state file */
+    for (gsize i=0; i < grouplength; i++)
+    {
+        auto group = std::string(groups[i]);
+        auto gp = get_prefix() + settings_type + " - ";
+        auto pos = group.find(gp);
+        if (pos == std::string::npos)
+            continue;
+
+        preset_names.push_back(group.substr(gp.size()));
+    }
+    // string array from the state file is no longer needed now.
+    g_strfreev (groups);
+
+    /* We want our settings to appear sorted alphabetically to the user */
+    std::sort(preset_names.begin(), preset_names.end());
+
+    /* Now add each preset to our global list */
+    presets_price.clear();
+
+    /* Start with the internally generated ones */
+    presets_price.push_back(create_int_no_preset());
+    //presets_price.push_back(create_int_gnc_exp_preset()); // Not Required
+
+    /* Then add all the ones we found in the state file */
+    for (auto preset_name : preset_names)
+    {
+        auto preset = std::make_shared<CsvPriceImpSettings>();
+        preset->m_settings_type = settings_type;
+        preset->m_name = preset_name;
+        preset->load();
+        presets_price.push_back(preset);
```

Comment:
This **may** need a cast to the parent class, but I suspect C++ will do this for you.